### DISCUSSION TITLE
Can not verify mails

### DISCUSSION THREAD

**Discussion by oliverFU** [krzyzanowskim/ObjectivePGP#201] [General]
- ObjectivePGP version:  0.14.0
- Platform (iOS/macOS):  iOS
- Installation method (CocoaPods/Carthage/...): build from master
- Project name: Letterbox

# Description
Hey,

thank you the project and your work is very helpful. I apologize for the (maybe silly) issue because you mention that cleartext signatures are not supported. But on the other hand CanaryMail uses your project....
I can not verify some (plain) mails from Thunderbird etc. (see example below). In general, I can not verify mails with attached signatures and I am wondering if I am doing a mistake (e.g. selecting the wrong plain text or wrong conversation from String to Data) or if it is not possible in general.

Does someone has experience with verifying plain texts? 
I am sorry but I did not know where to answer the question. Maybe issues are not the right place but people may find it...

Thanks and best
Oliver

# Steps to reproduce
I just created a simple swift-playground example. But if you are interested I can create a testcase pullrequest (with asc-file etc).
 
Sample code to reproduce the issue:

```swift

let key = """
-----BEGIN PGP PUBLIC KEY BLOCK-----

mQGNBFp+5vsBDACuHCvqCBlUT1O+IIQ0LOWsA2l/UAa+7PHNHotZJ22BtR//fmkd
rIesPye2MeX+1R14m7tHt+Aw5xwc9t40xPD1Crbc2cnMaYJ2Siy5GBKpZh1Sr3jq
9AQiNzYe1l3yPvnRZ5M0zgc0ueyd+b61sr4KBu8PQ5BODPLW81afPBlBgVB0FDI2
k1d9q4+r+obVIs43Hy6vB4YkUOyx5Fuaftj75Q86HNk3ig6fcvnRnbEmz+XifGYz
J5T/x2sZTGhg4CBDTDmEzdY0SFf7qgz4DYPrImlVksz5q0AXc22VbxuzRsK74SYK
Nix4i7gjaUZz6vNW+9qlJxUV4oJzj21KHH9EDlL2ErM7FYs4kI+POPChcFKTeJ8H
4WxFBh67aHiIvHpo3f8pwitPCkk0UYU0KHcaHLgVv9R0vExBj7BDQI1Qf/20z/Fj
fNz6Xgx4Lw4yGzePMopgsP2QEiKXC34g4F3dnXB6kg1l05lKuP+NhZF3qj139Yox
lwkntfoQIhwJDUEAEQEAAbQnYWxpY2VAZW56ZXZhbG9zLmRlIDxhbGljZUBlbnpl
dmFsb3MuZGU+iQG3BBMBCAAhBQJafub7AhsDBBUICQoHCwkIBwMCAQQWAwIBAh4B
AheAAAoJEE272pm01/0opcUL/iB9C9tEi+zZJfGRFM5X/VG/xOEo4s/UKfbVF58J
KDbt1r4TrAGK+Nx5F3zk1kgeC6hHWW0NAoUG6d0b2Qx/mfR6E6DUUePWrJWyOzDM
PoFK+9+o6CSRxuNUMvZ7/HaEWSdVdXpP84Ku1YNRYxlzX/lW+L+AhptfExgzFAMu
m75f8fpThe3+lpj5XwYtSwtbqea8YXPwWVs38uEhyEdm98mPZjeChK607JQiMwV3
nr2WnK4I99vTm5FURDrPQngIbJQuBabeaWyZhRZNW1Es0yuW9A7gIeqioeGOuZkK
ZZctuDYAQbor33gxF9vbzSKnDw8d0W2XFGG3XwM7Z8Ht8vUn4s+7pMhT+u9rX55U
OhgDgHZGW4RCOrLzfWHZ13udjVCLQcH4TOXqt5KdRyVJY/5662uHhQARW0du1cbP
BldycQElvH3CLpjtmcCgWNnr2Ldjdyz0Vk9XvMp4pEjYy5zjknMtKCS312NbPjHF
Aqm1Pp+/P4PIaumevEoaLIQFfokBugQTAQgAJAUCWn7m+wIbAwQVCAkKBwsJCAcD
AgEEFgMCAQIeAQIXgAIZAQAKCRBNu9qZtNf9KF8RC/4wROMbXDCW2G6CGXK1gpQV
0gBJOeu4I0JGiFEZVBEIhRqWTqoF/cU5inBlACcr2/eNrWze68rYPkEEpeQRVrFY
G7F00k6QP/FrZTVcJTYYGzlC9NQX9HJXU89o6Ou7hNSKue5uuCIOEe7Xf9049eYI
etlwxwuvr9i3VmF9UqZVIeAAVfgEwgfHMeOP+XOSMQnGYx1g6rrfytScn7XBMCow
9/qSvKjI38evNTCMEtFbNejEL1O/81mBWuEBVqbqS0woiN7J4UMphJFda/Bg97wV
jP8uAhs+D9IUhvykKNEDoCfZZhqTGqyJCvzdZR0+M5X3SNB2yud4o2nA5iCx6Kgu
/iLGN85uwyEXtYUGQ/dQps2sWAOQO65CM0HP0Mb2Oz6jQ3oyq7jXw9ZWWtP4jZjL
ymMtkCTaO46hOy6zQ3+8d25A7HX6qhZTFhPcn3ckiHYsE25kDGE6dMBVEw47qUZS
hFjM5QLUDdkJt1/z26EeHCVpqEw9A2ngmoJ9dX3UanW5AY0EWn7m+wEMAJ/uwYJs
yYM7kTIgAkYu6PhUrksvWbucXm98zSRxVsJigEklaipwuC4bddZkpgSOHg+TAlsS
XXTFPNaoHjiHZp7QE5NamWZ9wXu3Xv7M4bdN5h+MaIpvWITQhWTIantqqTYPv2ff
WIfCbfi5P+3VzBoT94o3CRgDuBiBThmflkL5OoV5mw3qkCUhFJAq279hm/wtSrLv
Iyu8pE06b/9vI4Q19cIu9wM2Pn+A4lsf3F7rHoC/+jlNy0EiUipC97Ln8b/0lHaS
W2zzx0j/MKPbqEwEvM6+MK0VxEPuZ89N3OIh3zMWc9rvlS11npGW3Cnxi83sWJsK
3xLKFG6LD5Ba8Q/tbwjAnJ/oJIGHv6U2k48ZkrgVoAcNxpBielFZ5pZXdKyKiPv8
RjSxLc/yic0NdqUn1hYNTcKp4EYXiCMBFPQ1u7sTG6bcRUrDI/6aceB11UZEUKdZ
GhQI/zxxI1MEoc8dirR4QLhKUukRUYh7ExuNl3o5KnSrWWs2EHjAmwM5ZwARAQAB
iQM+BBgBCAGoBQJafub7AhsMwN0gBBkBCAAGBQJafub7AAoJEKmH5Wf4h8YiqfEL
/iRJ0n+4xh0U+GQvgRIkyIE+OyxjNhrJr/CmrSQqDUKX24lYBEUlT9kqwXaJx/FN
JxZXFCCjlqxMp0i1YSGnR/CoJf5XwpfaJYoUWn6G4mYcEB17wY22uDDxICgw2/iw
2LcwFDeD96Aou17V3PIziLql+/Nm8eHdjtz8Fb9Q0kBe9KCc8CJdeE3BPde/6716
ZtF1CLqOguJk5jG9ZaZ5g8nrGIMIjw1tNHLTm8HZREvKmGHPnIPUccBSadIxk5i5
P3L7dx2yMrXuWvNT4wqGDeg4Eoc27vsUhJ3eM92ibV3/xFs30g68ra3edLIA5Ef0
cggL9mxFnTcoZd8HVmZ0DbHIOMbHailgj/kFLVbb2KA7EQwxmoX5ds1m/SCLpAbr
X2hnPb8zrt0DFM1pstav/alMkejAvVbuATXx+O1S1ZRQ6vdjRbOFVvFHwI3B94lX
ue8MyhNeMPv+JvRqjMsG5lEnxF71b6UwzWPwoDUNixdzmuTIm+LvpkdBl7yOHqmx
uwAKCRBNu9qZtNf9KOSwC/9UWuXrPaJdW+HcPXGGXz0rIeLxfs4R48B66ErflToz
qR5EuYY+3ZTD0BrBkzSmAxeBb8fR6YojnA9bKQJ+DuzkuZ1UCS/5zd+n4xmi/WYX
AerjxSG8L5TYf0CI3gFMB0YGgIio1lskOuZxhzQ/lHxPZ8tNfY8LrtFyTx4mNKzh
IVM1x0Tz+Cv7I6ns5NrI/QhQK6KLTnEG/6G6imXzcf/U+/MDidXCHuxojxSWjgJJ
4TKG67eicFTjetRH6JTN/44MyVWvMOEciE5Bdyo6Zd/qLUNS+p9gaCQNcB+/5Qsx
YR28lTV/aa0kNgsoStm0DfjFH3eXCk4Ct2AT62wmeHw3i9M1i01mjT1Bqpyd0ODe
9zaE/ETRa7vdD9BwIGDzjBKYRYdU9+jjwoO8yGLkdfnhbuJs6TYqzlNTAGHlUHMN
ZUJueWJ5Bh56w64pDCzpNmaplhsPvJPel77Me06565z0muLn/w2waOkndMaK3Rm0
gXhCQhUXVfrkXAAQg+jrg04=
=XzWc
-----END PGP PUBLIC KEY BLOCK-----
"""

var text = """
--jGY9eHsM08PgfVXXPT2f3AZCPJW4IwZj4
Content-Type: multipart/mixed; boundary="dibSubdyFVMsrRSzHsR83Pehjj4u2PlX3";
protected-headers="v1"
From: alice <alice@enzevalos.de>
To: Bob <bob@enzevalos.de>
Message-ID: <310eef89-5be5-6018-3ded-a9cda76a0342@enzevalos.de>
Subject: signed from Thunderbird

--dibSubdyFVMsrRSzHsR83Pehjj4u2PlX3
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable
Content-Language: en-US

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque dapibus
id diam ac volutpat. Sed quis cursus ante. Vestibulum eget gravida
felis. Nullam accumsan diam quis sem ornare lacinia. Aenean risus risus,
maximus quis faucibus et, maximus at nunc. Duis pharetra augue libero,
et congue diam varius eget. Nullam efficitur ex purus, non accumsan
tellus laoreet hendrerit. Suspendisse gravida interdum eros, eu
venenatis ante suscipit nec. Class aptent taciti sociosqu ad litora
torquent per conubia nostra, per inceptos himenaeos. Praesent
pellentesque cursus sem, non ornare nunc commodo vel. Praesent sed magna
at ligula ultricies sagittis malesuada non est. Nam maximus varius
mauris. Etiam dignissim congue ligula eu porta. Nunc rutrum nisl id
mauris efficitur ultrices. Maecenas sit amet velit ac mauris consequat
sagittis at et lorem.



--dibSubdyFVMsrRSzHsR83Pehjj4u2PlX3--


"""

let sig = """
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEEJRxNIIPfYdSrprTDTbvambTX/SgFAlvZjsUACgkQTbvambTX
/Sj26gv/VN0k/cDybI5RUU/qCIxcivOy3qM/9AauHjQQ/RJWlBAsVy/JU04xS+Qi
DdLmXP9LJa8np5T9B/rVJZrVbv5FjidsY6UsPbcqRcTZGe/X8ikscpRqUlYqkeAI
F4u2tKQiSaxbB4J9Mj4WindG3y3kcRUoTizFv7I7Zl6rLc2WVayLwFaA79pDv3M9
NEgwv5i7gE48dppEJLUBeXtpK00N5/BxJ0w7L0hirQZFo7FMTkDgCbYA+9seDfS4
lx4aHYjB+BUsV1n7dO6bdumnutP3eK1awk2Yyq4EJHizLNd/SsWVltOxukRqrUAp
qFlwclAEOxlWs0uLckMU6jZnhk4nUC8S20MLznNN4ASBjdo4mRIZFHIq9X8qLikl
uUvxM1Tj3+Vx/6JvHI3kxrq68QARkvTKOMOcLnPUOlQgEvUe9JZvJfSdm2xhqVa+
vBWUTsso1k+G40NdpH72SlTxFdIOlg5GX4MVVXzDyt+t2z+PVZFOPhMyp25h7efS
JPEF5Jke
=SZPp
-----END PGP SIGNATURE-----

"""
let binSig = try Armor.readArmored(sig)
let binKey = try Armor.readArmored(key)
let keys = try ObjectivePGP.readKeys(from: binKey)

if let data = text.data(using: .utf8) {
    do {
        try ObjectivePGP.verify(data, withSignature: binSig, using: keys, passphraseForKey: nil)
        print("Verified!")
    }
    catch {
        print(error)
    }
}
else {
    print("No data")
}

```

Plain mail:

Return-Path: <alice@enzevalos.de>
Delivered-To: bob@enzevalos.de
To: Bob <bob@enzevalos.de>
From: alice <alice@enzevalos.de>
Subject: signed from Thunderbird
Openpgp: preference=signencrypt
Message-ID: <310eef89-5be5-6018-3ded-a9cda76a0342@enzevalos.de>
Date: Wed, 31 Oct 2018 12:15:17 +0100
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:52.0)
 Gecko/20100101 Thunderbird/52.9.1
MIME-Version: 1.0
Content-Type: multipart/signed; micalg=pgp-sha512;
 protocol="application/pgp-signature";
 boundary="jGY9eHsM08PgfVXXPT2f3AZCPJW4IwZj4X" 

--jGY9eHsM08PgfVXXPT2f3AZCPJW4IwZj4X
Content-Type: multipart/mixed; boundary="dibSubdyFVMsrRSzHsR83Pehjj4u2PlX3";
 protected-headers="v1"
From: alice <alice@enzevalos.de>
To: Bob <bob@enzevalos.de>
Message-ID: <310eef89-5be5-6018-3ded-a9cda76a0342@enzevalos.de>
Subject: signed from Thunderbird

--dibSubdyFVMsrRSzHsR83Pehjj4u2PlX3
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable
Content-Language: en-US

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque dapibus
id diam ac volutpat. Sed quis cursus ante. Vestibulum eget gravida
felis. Nullam accumsan diam quis sem ornare lacinia. Aenean risus risus,
maximus quis faucibus et, maximus at nunc. Duis pharetra augue libero,
et congue diam varius eget. Nullam efficitur ex purus, non accumsan
tellus laoreet hendrerit. Suspendisse gravida interdum eros, eu
venenatis ante suscipit nec. Class aptent taciti sociosqu ad litora
torquent per conubia nostra, per inceptos himenaeos. Praesent
pellentesque cursus sem, non ornare nunc commodo vel. Praesent sed magna
at ligula ultricies sagittis malesuada non est. Nam maximus varius
mauris. Etiam dignissim congue ligula eu porta. Nunc rutrum nisl id
mauris efficitur ultrices. Maecenas sit amet velit ac mauris consequat
sagittis at et lorem.

--dibSubdyFVMsrRSzHsR83Pehjj4u2PlX3--

--jGY9eHsM08PgfVXXPT2f3AZCPJW4IwZj4X
Content-Type: application/pgp-signature; name="signature.asc"
Content-Description: OpenPGP digital signature
Content-Disposition: attachment; filename="signature.asc"

-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEEJRxNIIPfYdSrprTDTbvambTX/SgFAlvZjsUACgkQTbvambTX
/Sj26gv/VN0k/cDybI5RUU/qCIxcivOy3qM/9AauHjQQ/RJWlBAsVy/JU04xS+Qi
DdLmXP9LJa8np5T9B/rVJZrVbv5FjidsY6UsPbcqRcTZGe/X8ikscpRqUlYqkeAI
F4u2tKQiSaxbB4J9Mj4WindG3y3kcRUoTizFv7I7Zl6rLc2WVayLwFaA79pDv3M9
NEgwv5i7gE48dppEJLUBeXtpK00N5/BxJ0w7L0hirQZFo7FMTkDgCbYA+9seDfS4
lx4aHYjB+BUsV1n7dO6bdumnutP3eK1awk2Yyq4EJHizLNd/SsWVltOxukRqrUAp
qFlwclAEOxlWs0uLckMU6jZnhk4nUC8S20MLznNN4ASBjdo4mRIZFHIq9X8qLikl
uUvxM1Tj3+Vx/6JvHI3kxrq68QARkvTKOMOcLnPUOlQgEvUe9JZvJfSdm2xhqVa+
vBWUTsso1k+G40NdpH72SlTxFdIOlg5GX4MVVXzDyt+t2z+PVZFOPhMyp25h7efS
JPEF5Jke
=SZPp
-----END PGP SIGNATURE-----

--jGY9eHsM08PgfVXXPT2f3AZCPJW4IwZj4X--

**Comment by krzyzanowskim** [comment]
> I can create a testcase pullrequest (with asc-file etc)

That's the best way to get me on the task ;)

**Comment by oliverFU** [comment]
Hey Marcin,

do you think that ObjectivePGP should support this type of message or not? 

Thanks and best
Oliver
